### PR TITLE
Fix options passing in fetchJsonLd

### PR DESCRIPTION
**Pull request by benfavre**
| Bug fix?      | yes : Fixes non fresh parameters sent to httpClient
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tickets       | fixes https://github.com/api-platform/admin/issues/275
| License       | MIT

When implementing Authorization and proceeding to logging in, the httpClient sends out request without the Authorization header set, event after following all the recommended steps.
This is due to how the options are passed to the httpClient.

### CODE REVIEW

**Review by dunglas — APPROVED** [review]

**Comment by dunglas** [comment]
@benfavre the test failure looks related to this change. Can you take a look?

File: src/hydra/fetchResource.ts [hunk 1]
```diff
@@ -7,7 +7,7 @@ export default (
 ): Promise<any> => {
   return fetchJsonLd(
     resourceUrl,
-    Object.assign({ itemsPerPage: 0 }, options)
+    { itemsPerPage: 0 , ...options}
```
Comment (dunglas) [inline src/hydra/fetchResource.ts:10]: Actually, passing `itemsPerPage` looks like an (existing) bug. It should be added to the URL's query parameters (user the `URL` class), not passed as an option to fetch. It's why tests fail.

**Review by dunglas — CHANGES_REQUESTED** [review]

File: src/hydra/fetchResource.ts (hunk 1 above)
Comment (benfavre) [reply src/hydra/fetchResource.ts]: Seemlingy simply removing it here is fine then ? I've update the PR.
Thanks.

**Comment by alanpoulain** [comment]
Why would it solve the issue https://github.com/api-platform/admin/issues/275?

**Comment by benfavre** [comment]
> Why would it solve the issue [api-platform/admin#275](https://github.com/api-platform/admin/issues/275)?
"There is a Object.assign() call used instead of what could be object destructuring or plain passing."

When using Object.assign() in a function's arguments what happens is that the newly created object has an embedded reference to the original copy of "options".
What we actually want is to pass in the current value of "options" everytime.
Currently "options" is only updated later, probably in the nextTick, this is why it works most of the time but not always.

The best solution to this would be to change how the httpClient is selected, and use the httpClient supplied to the AdminGuesser component rather than using this embedded one.

**Comment by alanpoulain** [comment]
For the second part, it can be done in another PR.

I'm not sure to understand well what you are describing. If we refer to https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/assign, it says:
> It uses [[Get]] on the source and [[Set]] on the target, so it will invoke getters and setters. Therefore it assigns properties, versus copying or defining new properties.

So yes, the assigned properties can be references, but it shouldn't be an issue since we shouldn't change the properties by reference.

**Comment by benfavre** [comment]
That wasn't actually a usefull answer.

The reality is that out of the box the Authorization header parsing doesn't like having it's value set to "Bearer null"
image

Here : https://api-platform.com/docs/admin/authentication-support/ you recommend:
```js
const fetchHeaders = { Authorization: `Bearer ${window.localStorage.getItem("token")}` };
````

Calling a non protected route with a null Bearer still attemps authentification, so we end up with issues when loading the login page:
image


To overcome this issue I have set "fetchHeaders" to be a closure.
```js
const fetchHeaders = localStorage.getItem(config.tokenStorageKey) ?
    { Authorization: `Bearer ${localStorage.getItem()}`} :
    {};

const apiDocumentationParser = entrypoint => parseHydraDocumentation(entrypoint, { headers: new Headers(fetchHeaders)})
```

My fix allows fetchHeaders to be a closure so I can overcome the issue of having a "Bearer null"

**Comment by dunglas** [comment]
Thanks for the detailed explanation!

I prefer your patch that the change in `security.yaml`. Passing `Bearer null` is a bug and should be avoided. We also need to change the docs to use a closure as you suggest.

**Comment by nckenn** [comment]
I don't know if this a bug on your side, or maybe bug on react admin.

The token is not updating when successfully login:
```
const fetchHeaders = { 'Authorization': `Bearer ${window.localStorage.getItem("token")}` };
const fetchHydra = (url, options = {}) => baseFetchHydra(url, {
    ...options,
    headers: new Headers(fetchHeaders),
});
```
The picture below is my first login, and the token is set correctly without error, and then i logout without refreshing the page.


And i login again, but as you can see the token is not changed, it retained the previously stored token.


for your reference i save the token on my first login and second login which is saved on localstorage after successfully login :

1st Login Token
```
eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzI1NiJ9.eyJpYXQiOjE2MDI2MDQ4NzMsImV4cCI6MTYwMjY5MTI3Mywicm9sZXMiOlsiUk9MRV9TVVBFUl9BRE1JTiJdLCJ1c2VybmFtZSI6InN1cGVyYWRtaW4ifQ.MIh68XPuNKzaeanlge5diy2D_eKVsCogOfGYkkr46VVfzH5ZUPNUD0nUJYH9cLuhsYIkAXAuo2IhPxp1l807MadRgSzpL7icm7Hh2ZqpwtLO7ilR4HUCwpY9KWN4xiGD9m2l33SLs58KrKTOOVhnOdpeQV6CCBbuCZSvRPStaGzdwUun-WDYYWZpaN4vLx-bc8GVJ2dQqqWahJ-wGbL098fb_en4k8ts4POs35t1B5HzllMYDZHxQtT62gvVPownA8BnrPE26J4hVFwavO6WSTeTZgcpppwsKsuGhg4xPiTX4nAR84KAk7J5w24dKQdeJOdD9MT78FRGIuDt-v-HVg1nXPvzHcrEnd2piLjeNpYSt8B4_JvQhOc1Fw_Px2AZhjzQ8QyauFOxKm6NdcFbz35xfBO7Ws5JyFAldH6coQzL9gcT6TVo44F6Jgc_GOjkOsWdV3iwuClENdOfcO_gYm24OQvIk7W4ifd1CgT2Xwv-6iDzMnHA-ZhFMeajn_JMO8p1mO9_QxZGZrR19l8TAAk5C4saNzL4srE91WhtiPoEw2xT4mQP9cknLxaT1UV0oFouWmvEsxdHHalZKanU0cMlkj3QrDANRMEkDMYsae29bRO5xxqTa2lhC38dOpnreT7b_zKWccvXjEnWuHRSTiWndUGgRIgjP3ztadWWAhQ
```
2nd Login Token 
```

eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzI1NiJ9.eyJpYXQiOjE2MDI2MDQ5NzAsImV4cCI6MTYwMjY5MTM3MCwicm9sZXMiOlsiUk9MRV9TVVBFUl9BRE1JTiJdLCJ1c2VybmFtZSI6InN1cGVyYWRtaW4ifQ.OQDHQLtjRVBYtc1AcPAVkubNeRcM7F1K9PhIyNF_S0Jrr4t7cK3GnXItk40-IgZlNlwHfNbSjn5ngTKoObTK0ajknBCqMPWjJbhZETQMoS-mX2frti_NIj0fKJeV_5nsrQlxpxe4DB28Ob8bKBAQURZPQV1GUdxyb6FTEpX1d3IX-0iZV59vZ2QkWyxlrkf5C69v9bbUGYsqAhOh5Y5gVEde0k5JlRsQcUetv8lnLshaY2YrdxrzkEw0V_R3B1gibPnuj4IQZ693YdkeqFldGicV2Q4WIknz96ch68Ukd87Dv4mSzkA0HcUWUdFRDKa17jHaxhq6gKEaPwXsVt1zFdvcPnYwUCS6E7paA1_VS8X_gENoLIgu-vju65-_7moShcNhyDg5s9foxzJu3N05FqL-I_8bkwIXF9flUot-WGXY4k-Nj7okpbk3SskwqjDR6OrepEmAoKNCTw1ZEcIYgqq9Z2301WTt_il3lQHkXAb0ACCwDD2IxfN5EBefOAgtXIAEYAImf_rNlUfQU5ZpicMVJzHYQgpYGZLEdA1brcAjIV8kjWboOKr5cj9mlFYUN7-F_UL5qv1lz-CEtbGNF0kTxGyCJbuJm_ElTzTw4PKlyMe3rK1VFo_SplZ9DTB1qjZxkt0HdFlO3X-_nXU4_4aHGxJhQ-lBhPup7ZAwM5M
```


*My remedy for this is to reload page after successfully login.

**Comment by benfavre** [comment]
@nckenn this ticket adresses this exact issue, that's the point :D
The bug is in this repo and is fixed by this PR ... We are just waiting for it to be merged

**Comment by dunglas** [comment]
@nckenn could you confirm if the PR fixes the bug on your projet? Thanks!

**Comment by alanpoulain** [comment]
@benfavre Please follow this PR to fix your issue: https://github.com/api-platform/docs/pull/1184

**Comment by alanpoulain** [comment]
Should be solved with https://github.com/api-platform/docs/pull/1228.

**Pull request closed without merging**